File: src/core/scorer.py
"""Compute a deterministic Slop Score Index using local heuristics."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from ..filters.metadata import match_ad_layout_token, match_sponsored_sentinel
from ..filters.typography import has_typography_noise
from .parser import extract_card_strings, normalize_content_card


EM_DASH = "\u2014"
LIST_MARKER_RE = re.compile(r"^(?P<lead>\s*(?:[-*+]\s+|\d+\.\s+))(?P<body>.+)$")
SERIAL_SPLIT_RE = re.compile(r"\s*—\s*")
TRANSITION_PREFIXES = (
    "and",
    "at least",
    "because",
    "but",
    "especially",
    "even though",
    "for now",
    "instead",
    "just",
    "maybe",
    "perhaps",
    "rather",
    "so",
    "still",
    "then",
    "though",
    "while",
    "yet",
)
PARENTHETICAL_PREFIXES = (
    "as ",
    "which ",
    "who ",
    "whose ",
    "where ",
    "when ",
    "while ",
    "with ",
    "without ",
)
COLON_CUE_RE = re.compile(
    r"\b(is|are|was|were|means|meant|remains|remained|looks|looked|sounds|sounded|feels|felt|goal|result|plan|rule)\b",
    re.IGNORECASE,
)
ACTION_PREFIXES = (
    "by ",
    "just ",
    "paste ",
    "review ",
    "run ",
    "set up ",
    "to ",
    "use ",
)
SHORT_LABEL_TERMS = frozenset(
    {
        "guide",
        "logic",
        "priority",
        "reference",
        "sheet",
        "status",
        "tab",
        "tiers",
        "values",
    }
)

DEFAULT_THRESHOLDS = {
    "human": 25,
    "suspect": 60,
}


@dataclass(frozen=True)
class ScoreBreakdown:
    score: int
    bucket: str
    matched_rules: list[str]

@dataclass(frozen=True)
class ScoreResult:
    score: int
    bucket: str
    matched_rules: list[str]
    breakdown: dict[str, int]


def load_rules(path: str | Path) -> dict:
    return json.loads(Path(path).read_text(encoding="utf-8"))


def _word_count(text: str) -> int:
    return len(text.split())


def _looks_parenthetical(middle: str) -> bool:
    lowered = middle.strip().lower()
    return (
        2 <= _word_count(middle) <= 14
        and not middle.endswith((".", "!", "?"))
        and any(lowered.startswith(prefix) for prefix in PARENTHETICAL_PREFIXES)
    )


def _looks_descriptor(label: str) -> bool:
    if _word_count(label) > 6:
        return False
    lowered = f" {label.lower()} "
    return not any(token in lowered for token in (" is ", " are ", " was ", " were ", " has ", " have "))


def _should_use_comma(right: str) -> bool:
    lowered = right.strip().lower()
    return any(lowered.startswith(prefix) for prefix in TRANSITION_PREFIXES)


def _should_use_colon(left: str, right: str) -> bool:
    right_lower = right.strip().lower()
    if "=" in right and 1 <= _word_count(left) <= 5:
        return True
    if "=" in left and 1 <= _word_count(right) <= 4:
        return True
    if any(right_lower.startswith(prefix) for prefix in ACTION_PREFIXES):
        return bool(COLON_CUE_RE.search(left))
    if right_lower.startswith(("for example", "namely", "specifically")):
        return True
    return False


def _is_title_fragment(text: str) -> bool:
    words = [word for word in re.split(r"\s+", text.strip()) if word]
    if not words or len(words) > 8:
        return False
    for word in words:
        stripped = word.strip("()[]{}:;,.'\"")
        if not stripped:
            continue
        if stripped[0].isdigit():
            continue
        if stripped[0].isupper():
            continue
        if stripped.lower() in {"and", "of", "the", "for", "in", "to", "&"}:
            continue
        return False
    return True


def _looks_heading_style(left: str, right: str, line: str) -> bool:
    stripped = line.strip()
    if not stripped or stripped != line:
        return False
    if stripped.endswith((".", "!", "?", ":", ";")):
        return False
    if "\t" in stripped or stripped.startswith(("-", "*", "+")):
        return False
    if left.lower().startswith("step ") and 1 <= _word_count(right) <= 10:
        return True
    if _is_title_fragment(left) and _is_title_fragment(right):
        return True
    left_words = left.split()
    if 1 <= len(left_words) <= 5 and left_words[-1].lower() in SHORT_LABEL_TERMS:
        return True
    return False


def _score_line(
    line: str,
    weights: dict[str, int],
    *,
    allow_title_fragment: bool = True,
) -> tuple[int, list[str], dict[str, int]]:
    score = 0
    matches: list[str] = []
    breakdown: dict[str, int] = {}
    count = line.count(EM_DASH)

    if count == 0:
        if not allow_title_fragment:
            return score, matches, breakdown
        if _is_title_fragment(line):
            penalty = int(weights.get("title_fragment", 0))
            if penalty:
                score += penalty
                matches.append("title_fragment")
                breakdown["title_fragment"] = penalty
        return score, matches, breakdown

    parts = [part.strip() for part in SERIAL_SPLIT_RE.split(line)]
    if len(parts) == 3 and _looks_parenthetical(parts[1]):
        penalty = int(weights.get("parenthetical_dash", 0))
        score += penalty
        matches.append("dash_parenthetical")
        breakdown["dash_parenthetical"] = penalty
        return score, matches, breakdown

    if len(parts) >= 3 and not any(not part for part in parts):
        if all(_word_count(part) <= 10 for part in parts[:-1]):
            penalty = int(weights.get("serial_dash", 0)) + max(0, len(parts) - 3) * int(
                weights.get("serial_dash_step", 0)
            )
            score += penalty
            matches.append("dash_serial")
            breakdown["dash_serial"] = penalty
            return score, matches, breakdown

    if count == 1:
        list_match = LIST_MARKER_RE.match(line)
        if list_match is not None:
            body = list_match.group("body")
            if body.count(EM_DASH) == 1:
                label, detail = [part.strip() for part in body.split(EM_DASH, 1)]
                if _looks_descriptor(label):
                    penalty = int(weights.get("list_descriptor_dash", 0))
                    score += penalty
                    matches.append("dash_list_descriptor")
                    breakdown["dash_list_descriptor"] = penalty
                    return score, matches, breakdown

        left, right = [part.strip() for part in line.split(EM_DASH, 1)]
        if _looks_heading_style(left, right, line):
            penalty = int(weights.get("heading_style_dash", 0))
            score += penalty
            matches.append("dash_heading_style")
            breakdown["dash_heading_style"] = penalty
            return score, matches, breakdown
        if _should_use_colon(left, right):
            penalty = int(weights.get("colon_candidate_dash", 0))
            score += penalty
            matches.append("dash_colon_candidate")
            breakdown["dash_colon_candidate"] = penalty
            return score, matches, breakdown
        if _should_use_comma(right):
            penalty = int(weights.get("comma_candidate_dash", 0))
            score += penalty
            matches.append("dash_comma_candidate")
            breakdown["dash_comma_candidate"] = penalty
            return score, matches, breakdown

        penalty = int(weights.get("single_dash_fallback", 0))
        score += penalty
        matches.append("dash_fallback")
        breakdown["dash_fallback"] = penalty
        return score, matches, breakdown

    penalty = int(weights.get("multi_dash_fallback", 0)) + max(0, count - 2) * int(weights.get("multi_dash_step", 0))
    score += penalty
    matches.append("dash_multi_fallback")
    breakdown["dash_multi_fallback"] = penalty
    return score, matches, breakdown


def _score_single_field(
    text: str,
    rules: dict,
    field: str,
) -> tuple[int, list[str], dict[str, int]]:
    weights = rules.get("weights", {})
    headline_like_pattern = rules.get("regex_patterns", {}).get("headline_like")
    score = 0
    matches: list[str] = []
    breakdown: dict[str, int] = {}

    if not text:
        return score, matches, breakdown

    ad_token = match_ad_layout_token(text)
    if ad_token is not None:
        penalty = int(weights.get("exact_ad_layout_token", 100))
        score += penalty
        matches.append(f"layout:{ad_token}")
        breakdown[f"layout:{ad_token}"] = penalty

    sentinel = match_sponsored_sentinel(text, rules.get("sentinel_words", []))
    if sentinel is not None:
        penalty = int(weights.get("sponsored_sentinel", 35))
        score += penalty
        matches.append(f"sentinel:{sentinel}")
        breakdown[f"sentinel:{sentinel}"] = penalty

    if has_typography_noise(text):
        penalty = int(weights.get("typography_noise", 0))
        if penalty:
            score += penalty
            matches.append("typography_noise")
            breakdown["typography_noise"] = breakdown.get("typography_noise", 0) + penalty

    if headline_like_pattern is not None:
        stripped = text.strip().strip("\"'“”‘’()[]{}")
        if stripped and re.search(headline_like_pattern, stripped):
            penalty = int(weights.get("headline_like", 0))
            if penalty:
                score += penalty
                matches.append("headline_like")
                breakdown["headline_like"] = breakdown.get("headline_like", 0) + penalty

    line_score, line_matches, line_breakdown = _score_line(
        text,
        weights,
        allow_title_fragment=field != "publisher",
    )
    score += line_score
    matches.extend(line_matches)
    for key, value in line_breakdown.items():
        breakdown[key] = breakdown.get(key, 0) + value

    return score, matches, breakdown


def _score_text_fields(fields: list[tuple[str, str]], rules: dict) -> tuple[ScoreBreakdown, dict[str, int]]:
    score = 0
    matches: list[str] = []
    breakdown: dict[str, int] = {}

    for field, text in fields:
        field_score, field_matches, field_breakdown = _score_single_field(text, rules, field)
        score += field_score
        matches.extend(field_matches)
        for key, value in field_breakdown.items():
            breakdown[key] = breakdown.get(key, 0) + value

    score = max(0, min(100, score))
    thresholds = rules.get("thresholds", DEFAULT_THRESHOLDS)
    human = int(thresholds.get("human", DEFAULT_THRESHOLDS["human"]))
    suspect = int(thresholds.get("suspect", DEFAULT_THRESHOLDS["suspect"]))
    if score < human:
        bucket = "Human"
    elif score < suspect:
        bucket = "Suspect (Highlight)"
    else:
        bucket = "High Probability Slop (Block)"
    return ScoreBreakdown(score=score, bucket=bucket, matched_rules=matches), breakdown


def score_text(text: str, rules: dict | None = None) -> ScoreResult:
    """Score a raw text blob."""
    rules = rules or {}
    score_breakdown, breakdown = _score_text_fields([("text", text)], rules)
    return ScoreResult(
        score=score_breakdown.score,
        bucket=score_breakdown.bucket,
        matched_rules=score_breakdown.matched_rules,
        breakdown=breakdown,
    )


def score_content_card(card: Any, rules: dict | None = None) -> ScoreResult:
    """Score a single content card deterministically."""
    rules = rules or {}
    normalized = normalize_content_card(card)
    fields = [
        ("title", normalized.title),
        ("excerpt", normalized.excerpt),
        ("publisher", normalized.publisher),
    ]
    if not any(text for _, text in fields):
        fields = [("text", text) for text in extract_card_strings(normalized)]
    score_breakdown, breakdown = _score_text_fields(fields, rules)
    return ScoreResult(
        score=score_breakdown.score,
        bucket=score_breakdown.bucket,
        matched_rules=score_breakdown.matched_rules,
        breakdown=breakdown,
    )

File: src/extension/scorer.js
const EM_DASH = "\u2014";
const LIST_MARKER_RE = /^(?<lead>\s*(?:[-*+]\s+|\d+\.\s+))(?<body>.+)$/;
const SERIAL_SPLIT_RE = /\s*—\s*/;
const TRANSITION_PREFIXES = [
  "and",
  "at least",
  "because",
  "but",
  "especially",
  "even though",
  "for now",
  "instead",
  "just",
  "maybe",
  "perhaps",
  "rather",
  "so",
  "still",
  "then",
  "though",
  "while",
  "yet",
];
const PARENTHETICAL_PREFIXES = [
  "as ",
  "which ",
  "who ",
  "whose ",
  "where ",
  "when ",
  "while ",
  "with ",
  "without ",
];
const COLON_CUE_RE = /\b(is|are|was|were|means|meant|remains|remained|looks|looked|sounds|sounded|feels|felt|goal|result|plan|rule)\b/i;
const ACTION_PREFIXES = ["by ", "just ", "paste ", "review ", "run ", "set up ", "to ", "use "];
const SHORT_LABEL_TERMS = new Set(["guide", "logic", "priority", "reference", "sheet", "status", "tab", "tiers", "values"]);
const SPONSORED_SENTINELS = ["sponsored", "promoted", "partner content", "native ad", "paid post"];
const DEFAULT_THRESHOLDS = { human: 25, suspect: 60 };

function wordCount(text) {
  const trimmed = String(text).trim();
  if (!trimmed) {
    return 0;
  }
  return trimmed.split(/\s+/).length;
}

function isTitleFragment(text) {
  const words = String(text)
    .trim()
    .split(/\s+/)
    .filter(Boolean);
  if (words.length === 0 || words.length > 8) {
    return false;
  }
  for (const word of words) {
    const stripped = word.replace(/^[()\[\]{}:;,.'"]+|[()\[\]{}:;,.'"]+$/g, "");
    if (!stripped) {
      continue;
    }
    if (/^\d/.test(stripped)) {
      continue;
    }
    if (/^[A-Z]/.test(stripped)) {
      continue;
    }
    if (["and", "of", "the", "for", "in", "to", "&"].includes(stripped.toLowerCase())) {
      continue;
    }
    return false;
  }
  return true;
}

function looksParenthetical(middle) {
  const lowered = String(middle).trim().toLowerCase();
  return (
    wordCount(middle) >= 2 &&
    wordCount(middle) <= 14 &&
    !/[.!?]$/.test(String(middle).trim()) &&
    PARENTHETICAL_PREFIXES.some((prefix) => lowered.startsWith(prefix))
  );
}

function looksDescriptor(label) {
  if (wordCount(label) > 6) {
    return false;
  }
  const lowered = ` ${String(label).toLowerCase()} `;
  return ![" is ", " are ", " was ", " were ", " has ", " have "].some((token) => lowered.includes(token));
}

function shouldUseComma(right) {
  const lowered = String(right).trim().toLowerCase();
  return TRANSITION_PREFIXES.some((prefix) => lowered.startsWith(prefix));
}

function shouldUseColon(left, right) {
  const rightLower = String(right).trim().toLowerCase();
  if (String(right).includes("=") && wordCount(left) >= 1 && wordCount(left) <= 5) {
    return true;
  }
  if (String(left).includes("=") && wordCount(right) >= 1 && wordCount(right) <= 4) {
    return true;
  }
  if (ACTION_PREFIXES.some((prefix) => rightLower.startsWith(prefix))) {
    return COLON_CUE_RE.test(left);
  }
  if (rightLower.startsWith("for example") || rightLower.startsWith("namely") || rightLower.startsWith("specifically")) {
    return true;
  }
  return false;
}

function looksHeadingStyle(left, right, line) {
  const stripped = String(line).trim();
  if (!stripped || stripped !== line) {
    return false;
  }
  if (/[.!?:;]$/.test(stripped)) {
    return false;
  }
  if (/\t/.test(stripped) || /^[-*+]/.test(stripped)) {
    return false;
  }
  if (String(left).toLowerCase().startsWith("step ") && wordCount(right) >= 1 && wordCount(right) <= 10) {
    return true;
  }
  if (isTitleFragment(left) && isTitleFragment(right)) {
    return true;
  }
  const leftWords = String(left).split(/\s+/);
  if (leftWords.length >= 1 && leftWords.length <= 5 && SHORT_LABEL_TERMS.has(leftWords[leftWords.length - 1].toLowerCase())) {
    return true;
  }
  return false;
}

function normalizeWhitespace(text) {
  return String(text).split(/\s+/).filter(Boolean).join(" ");
}

export function matchAdLayoutToken(text) {
  const normalized = normalizeWhitespace(text);
  for (const token of ["AmazonSponsored", "WalmartSponsored", "PocketSponsored", "SponsoredContent", "PromotedContent", "NativeAd"]) {
    if (normalized.includes(token)) {
      return token;
    }
  }
  return null;
}

export function matchSponsoredSentinel(text, sentinelWords = []) {
  const lowered = String(text).toLowerCase();
  const candidates = [...new Set([...SPONSORED_SENTINELS, ...sentinelWords].map((word) => String(word).toLowerCase()))];
  for (const token of candidates) {
    if (token && lowered.includes(token)) {
      return token;
    }
  }
  return null;
}

export function hasTypographyNoise(text) {
  return /[!?]{3,}|\.{4,}/.test(String(text));
}

export function extractStrings(payload) {
  const result = [];
  const walk = (value) => {
    if (value === null || value === undefined) {
      return;
    }
    if (typeof value === "string") {
      const text = value.trim();
      if (text) {
        result.push(text);
      }
      return;
    }
    if (Array.isArray(value)) {
      for (const item of value) {
        walk(item);
      }
      return;
    }
    if (typeof value === "object") {
      for (const item of Object.values(value)) {
        walk(item);
      }
    }
  };

  walk(payload);
  return result;
}

export function normalizeContentCard(card) {
  if (card && typeof card === "object" && !Array.isArray(card)) {
    return {
      title: String(card.title ?? "").trim(),
      excerpt: String(card.excerpt ?? "").trim(),
      publisher: String(card.publisher ?? "").trim(),
    };
  }
  const [title = "", excerpt = "", publisher = ""] = [...extractStrings(card), "", "", ""];
  return {
    title: String(title).trim(),
    excerpt: String(excerpt).trim(),
    publisher: String(publisher).trim(),
  };
}

export function extractCardStrings(card) {
  const normalized = normalizeContentCard(card);
  const values = [normalized.title, normalized.excerpt, normalized.publisher].filter(Boolean);
  if (values.length) {
    return values;
  }
  return extractStrings(card);
}

function scoreLine(line, weights, { allowTitleFragment = true } = {}) {
  let score = 0;
  const matches = [];
  const breakdown = {};
  const count = String(line).split(EM_DASH).length - 1;

  if (count === 0) {
    if (!allowTitleFragment) {
      return { score, matches, breakdown };
    }
    if (isTitleFragment(line)) {
      const penalty = Number(weights.title_fragment ?? 0);
      if (penalty) {
        score += penalty;
        matches.push("title_fragment");
        breakdown.title_fragment = penalty;
      }
    }
    return { score, matches, breakdown };
  }

  const parts = String(line).split(SERIAL_SPLIT_RE).map((part) => part.trim());
  if (parts.length === 3 && looksParenthetical(parts[1])) {
    const penalty = Number(weights.parenthetical_dash ?? 0);
    score += penalty;
    matches.push("dash_parenthetical");
    breakdown.dash_parenthetical = penalty;
    return { score, matches, breakdown };
  }

  if (parts.length >= 3 && !parts.some((part) => !part)) {
    if (parts.slice(0, -1).every((part) => wordCount(part) <= 10)) {
      const penalty = Number(weights.serial_dash ?? 0) + Math.max(0, parts.length - 3) * Number(weights.serial_dash_step ?? 0);
      score += penalty;
      matches.push("dash_serial");
      breakdown.dash_serial = penalty;
      return { score, matches, breakdown };
    }
  }

  if (count === 1) {
    const listMatch = String(line).match(LIST_MARKER_RE);
    if (listMatch) {
      const body = listMatch.groups?.body ?? "";
      if ((body.split(EM_DASH).length - 1) === 1) {
        const [label] = body.split(EM_DASH).map((part) => part.trim());
        if (looksDescriptor(label)) {
          const penalty = Number(weights.list_descriptor_dash ?? 0);
          score += penalty;
          matches.push("dash_list_descriptor");
          breakdown.dash_list_descriptor = penalty;
          return { score, matches, breakdown };
        }
      }
    }

    const [left, right] = String(line).split(EM_DASH).map((part) => part.trim());
    if (looksHeadingStyle(left, right, line)) {
      const penalty = Number(weights.heading_style_dash ?? 0);
      score += penalty;
      matches.push("dash_heading_style");
      breakdown.dash_heading_style = penalty;
      return { score, matches, breakdown };
    }
    if (shouldUseColon(left, right)) {
      const penalty = Number(weights.colon_candidate_dash ?? 0);
      score += penalty;
      matches.push("dash_colon_candidate");
      breakdown.dash_colon_candidate = penalty;
      return { score, matches, breakdown };
    }
    if (shouldUseComma(right)) {
      const penalty = Number(weights.comma_candidate_dash ?? 0);
      score += penalty;
      matches.push("dash_comma_candidate");
      breakdown.dash_comma_candidate = penalty;
      return { score, matches, breakdown };
    }

    const penalty = Number(weights.single_dash_fallback ?? 0);
    score += penalty;
    matches.push("dash_fallback");
    breakdown.dash_fallback = penalty;
    return { score, matches, breakdown };
  }

  const penalty = Number(weights.multi_dash_fallback ?? 0) + Math.max(0, count - 2) * Number(weights.multi_dash_step ?? 0);
  score += penalty;
  matches.push("dash_multi_fallback");
  breakdown.dash_multi_fallback = penalty;
  return { score, matches, breakdown };
}

function buildRegex(pattern) {
  if (!pattern) {
    return null;
  }
  return new RegExp(pattern);
}

function scoreSingleField(text, rules, field) {
  const weights = rules.weights ?? {};
  const headlineLikePattern = buildRegex(rules.regex_patterns?.headline_like);
  let score = 0;
  const matches = [];
  const breakdown = {};

  if (!text) {
    return { score, matches, breakdown };
  }

  const adToken = matchAdLayoutToken(text);
  if (adToken !== null) {
    const penalty = Number(weights.exact_ad_layout_token ?? 100);
    score += penalty;
    matches.push(`layout:${adToken}`);
    breakdown[`layout:${adToken}`] = penalty;
  }

  const sentinel = matchSponsoredSentinel(text, rules.sentinel_words ?? []);
  if (sentinel !== null) {
    const penalty = Number(weights.sponsored_sentinel ?? 35);
    score += penalty;
    matches.push(`sentinel:${sentinel}`);
    breakdown[`sentinel:${sentinel}`] = penalty;
  }

  if (hasTypographyNoise(text)) {
    const penalty = Number(weights.typography_noise ?? 0);
    if (penalty) {
      score += penalty;
      matches.push("typography_noise");
      breakdown.typography_noise = (breakdown.typography_noise ?? 0) + penalty;
    }
  }

  if (headlineLikePattern) {
    const stripped = String(text).trim().replace(/^[\"'“”‘’()\[\]{}]+|[\"'“”‘’()\[\]{}]+$/g, "");
    if (stripped && headlineLikePattern.test(stripped)) {
      const penalty = Number(weights.headline_like ?? 0);
      if (penalty) {
        score += penalty;
        matches.push("headline_like");
        breakdown.headline_like = (breakdown.headline_like ?? 0) + penalty;
      }
    }
  }

  const lineResult = scoreLine(text, weights, { allowTitleFragment: field !== "publisher" });
  score += lineResult.score;
  matches.push(...lineResult.matches);
  for (const [key, value] of Object.entries(lineResult.breakdown)) {
    breakdown[key] = (breakdown[key] ?? 0) + value;
  }

  return { score, matches, breakdown };
}

function scoreTextFields(fields, rules) {
  let score = 0;
  const matches = [];
  const breakdown = {};

  for (const [field, text] of fields) {
    const fieldResult = scoreSingleField(text, rules, field);
    score += fieldResult.score;
    matches.push(...fieldResult.matches);
    for (const [key, value] of Object.entries(fieldResult.breakdown)) {
      breakdown[key] = (breakdown[key] ?? 0) + value;
    }
  }

  score = Math.max(0, Math.min(100, score));
  const thresholds = rules.thresholds ?? DEFAULT_THRESHOLDS;
  const human = Number(thresholds.human ?? DEFAULT_THRESHOLDS.human);
  const suspect = Number(thresholds.suspect ?? DEFAULT_THRESHOLDS.suspect);
  const bucket = score < human ? "Human" : score < suspect ? "Suspect (Highlight)" : "High Probability Slop (Block)";
  return { score, bucket, matchedRules: matches, breakdown };
}

export function scoreText(text, rules = {}) {
  const result = scoreTextFields([["text", text]], rules);
  return result;
}

export function scoreContentCard(card, rules = {}) {
  const normalized = normalizeContentCard(card);
  const fields = [
    ["title", normalized.title],
    ["excerpt", normalized.excerpt],
    ["publisher", normalized.publisher],
  ];
  if (fields.some(([, text]) => text)) {
    return scoreTextFields(fields, rules);
  }
  return scoreTextFields(extractCardStrings(normalized).map((text) => ["text", text]), rules);
}
